plugin/buycourses/src/buy_course_plugin.class.php

```diff
@@ -2917,6 +3132,313 @@ public function getBuyCoursePluginPrice(Session $session)
+            return false;
+        }
+
+        $values = [
+            'coupon_id' => (int) $couponId,
+            'service_sale_id' => (int) $saleId,
+        ];
+
+        return Database::insert(self::TABLE_COUPON_SERVICE_SALE, $values);
+    }
+
+    /**
+     * Add a new coupon.
+     *
+     * @param int $coupon
+     *
+     * @return bool
+     */
+    public function addNewCoupon($coupon)
+    {
+        $couponId = $this->registerCoupon($coupon);
+        if ($couponId) {
+            if (isset($coupon['courses'])) {
+                foreach ($coupon['courses'] as $course) {
+                    $this->registerCouponItem($couponId, self::PRODUCT_TYPE_COURSE, $course);
+                }
+            }
+
+            if (isset($coupon['sessions'])) {
+                foreach ($coupon['sessions'] as $session) {
+                    $this->registerCouponItem($couponId, self::PRODUCT_TYPE_SESSION, $session);
+                }
+            }
+
+            if (isset($coupon['services'])) {
+                foreach ($coupon['services'] as $service) {
+                    $this->registerCouponService($couponId, $service);
+                }
+            }
+
+            return true;
+        } else {
+            Display::addFlash(
+                Display::return_message(
+                    $this->get_lang('CouponErrorInsert'),
+                    'error',
+                    false
+                )
+            );
+
+            return false;
+        }
+    }
+
+    /**
+     * Add a new coupon.
+     *
+     * @param array $coupon
+     *
+     * @return bool
+     */
+    public function updateCouponData($coupon)
+    {
+        $this->updateCoupon($coupon);
+        $this->deleteCouponItemsByCoupon(self::PRODUCT_TYPE_COURSE, $coupon['id']);
+        $this->deleteCouponItemsByCoupon(self::PRODUCT_TYPE_SESSION, $coupon['id']);
+        $this->deleteCouponServicesByCoupon($coupon['id']);
+
+        if (isset($coupon['courses'])) {
+            foreach ($coupon['courses'] as $course) {
+                $this->registerCouponItem($coupon['id'], self::PRODUCT_TYPE_COURSE, $course);
+            }
+        }
+
+        if (isset($coupon['sessions'])) {
+            foreach ($coupon['sessions'] as $session) {
+                $this->registerCouponItem($coupon['id'], self::PRODUCT_TYPE_SESSION, $session);
+            }
+        }
+
+        if (isset($coupon['services'])) {
+            foreach ($coupon['services'] as $service) {
+                $this->registerCouponService($coupon['id'], $service);
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Update coupons delivered.
+     *
+     * @param int $couponId The coupon ID
+     *
+     * @return bool
+     */
+    public function updateCouponDelivered($couponId)
+    {
+        $couponTable = Database::get_main_table(self::TABLE_COUPON);
+
+        $sql = "UPDATE $couponTable
+        SET delivered = delivered+1
+        WHERE
+            id = $couponId";
```

Comment:
Filtering should happen here on $couponId as it is the last point before the database. `(int) $couponId` would be enough.
Database::query() does not automatically filter params. Database::insert() and Database::update() do.